### ISSUE TITLE
Please take care of compile warnings

### ISSUE THREAD

**Issue by tarsius** [TheBB/spaceline#17]
Compiling causes many warnings. Could you please investigate these and where appropriate silence them. Thanks.

```
Compiling /home/jonas/.emacs.d/lib/spaceline/spaceline-config.el...
Compiling /home/jonas/.emacs.d/lib/spaceline/spaceline-segments.el...

In spaceline--segment-selection-info:
spaceline-segments.el:147:63:Warning: reference to free variable ‘evil-state’
spaceline-segments.el:150:58:Warning: reference to free variable
    ‘evil-visual-selection’

In spaceline--segment-hud:
spaceline-segments.el:161:18:Warning: reference to free variable
    ‘highlight-face’
spaceline-segments.el:161:33:Warning: reference to free variable
    ‘default-face’

In spaceline--segment-anzu:
spaceline-segments.el:172:14:Warning: reference to free variable ‘active’

In spaceline--segment-auto-compile:
spaceline-segments.el:177:14:Warning: reference to free variable ‘active’

In spaceline--segment-erc-track:
spaceline-segments.el:183:11:Warning: reference to free variable
    ‘erc-modified-channels-alist’

In spaceline--segment-org-pomodoro:
spaceline-segments.el:217:20:Warning: reference to free variable
    ‘org-pomodoro-mode-line’

In spaceline--segment-nyan-cat:
spaceline-segments.el:221:32:Warning: reference to free variable
    ‘default-face’

In spaceline--segment-evil-state:
spaceline-segments.el:301:32:Warning: reference to free variable ‘evil-state’

In end of data:
spaceline-segments.el:307:1:Warning: the following functions are not known to
    be defined: anzu--update-mode-line, mode-line-auto-compile-control,
    fancy-battery-default-mode-line, fancy-battery-powerline-face,
    nyan-create, window-numbering-get-number, eyebrowse--get,
    evil-state-property
Compiling /home/jonas/.emacs.d/lib/spaceline/spaceline.el...

In spaceline-highlight-face-evil-state:
spaceline.el:147:39:Warning: reference to free variable ‘evil-state’
spaceline.el:147:51:Warning: reference to free variable ‘evil-previous-state’
Done (Total of 3 files compiled, 1 skipped in 4 directories)
```

**Comment by TheBB** [comment]
I don't want to wrap every reference to a variable I don't control in a `boundp`. Most are implicitly wrapped in a `when (bound-and-true-p whatever-mode)` section wherein I access other variables from that package. What do you suggest, exactly?

**Comment by tarsius** [comment]
You can just *declare* that these variables and functions are defined:

```lisp
(declare-function foo-bar 'foo)
(defvar foo)
```

More worrisome are warnings like this:

```
spaceline-segments.el:161:18:Warning: reference to free variable
    ‘highlight-face’
```

I am guessing you are let-binding in some function and then using the value in another function. That second function and the byte-compiler do not know whether it is only ever called from a function that let-binds or not, and so the compiler complains. If that is the case, then you should define a properly prefixed variable instead:

```lisp
(defvar spaceline--default-face nil)
```

That makes the intention more clear and the compiler will be satisfied.

> What do you suggest, exactly?

When there are warnings then one doesn't know whether they are about a real issue, which could cause things to blow up, or if the maintainer just didn't bother suppressing irrelevant warnings. So the irrelevant ones should be suppressed (after making sure they are indeed irrelevant).

(I didn't submit a pr, because *I* don't know whether these warnings are relevant or not.)

**Comment by TheBB** [comment]
The first solution is fine for things in other packages. As for `active` and the others, I don't want to clutter the DSL by making the user prefix such a natural variable. I can probably work around it in the macro that generates the final segment code.

> When there are warnings then one doesn't know whether they are about a real issue, which could cause things to blow up, or if the maintainer just didn't bother suppressing irrelevant warnings.

It seems quite par for packages to generate compilation warnings. I've certainly never paid any heed to those from other packages I use.

**Comment by tarsius** [comment]
> As for active and the others, I don't want to clutter the DSL by making the user prefix such a natural variable.

After having looked at the documentation again I can understand now why you don't want to do that, and I agree.

> I can probably work around it in the macro that generates the final segment code.

I have done that for you now (#NN). It's not exactly pretty passing these values around, but I think it is better than to rely on the dynamic bindings.

I have also added the declarations for the symbols from optional dependencies.

**Comment by TheBB** [comment]
Closed by #19.
